Add bluetooth remote control via GPIO

AlexProgrammerDE opened the pull request:
Change-type: major

Hi everyone! With #83 becomes balena-sound to a multi speaker project. But i think a good speaker has its own on-board music and volume control buttons. Here is a example:

![buttons](https://user-images.githubusercontent.com/40795980/74084516-32f60d00-4a70-11ea-9bfe-1eda369ade29.jpg)

Roadmap: (At the top are the more important things. )

finished:
- [x] Find a bluetooth remote control service (Bluealsa itself: https://github.com/Arkq/bluez-alsa#configuration--usage)
- [x] Make this feature 100% optional (It should be able to deactivate it 100%.)
- [x] Rename bluetooth-audio to bluetooth 
- [x] Decide, which GPIO pins should be used ((Pin) Input: 11, 13 and 15 Output: 1, 17 and 4)
- [x] Delete buffer
- [x] Implement the bluetooth remote control service
- [x] Allow discovering via a button 
- [x] Add skipping in both directions
- [x] Write README info

bugs:

@tmigone and @chrisys i will maybe need help from your side. If i run into some problems i will let you know in the comment section below. 

Things i have to remember:
`echo "$(amixer -D bluealsa scontrols)" | sed -n " s,[^']*'\([^']*\).*,\1,p "`
`VOLUME="$(amixer -D bluealsa get "$DEVICE" | awk '$0~/%/{print $5}' | tr -d '[]%')"`
`VOLUME_PARSED="$(echo "${VOLUME}" | head -1)"`
https://askubuntu.com/questions/1113050/sending-pause-resume-playing-and-next-previous-track-bluetooth-commands
`dbus-send --system --dest=org.bluez --print-reply /org/bluez/hci0/dev_$MAC_PARSED org.bluez.MediaControl1.Play > /dev/null`

AlexProgrammerDE commented:
@tmigone i have a problem. Are there any ways of communicating between the multi room devices. I had an idea: Pressing buttons on the client device is acting like pressing on the server device itself.

chrisys commented:
@AlexProgrammerDE great idea here! I think this is going to be an awesome addition to the project.

AlexProgrammerDE commented:
@chrisys thanks! One advantage of my code is that it controls directly the volume on the phone it is connected to. It all works great, but has to be implemented.

AlexProgrammerDE commented:
@chrisys i found the problem with:
11.02.20 18:16:30 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 143 != 137
11.02.20 18:16:31 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 150 != 144
11.02.20 18:16:31 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 157 != 152
11.02.20 18:16:31 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 167 != 158
11.02.20 18:16:31 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 175 != 168
11.02.20 18:16:31 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 182 != 176
11.02.20 18:16:31 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 192 != 183
11.02.20 18:16:32 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 200 != 193
11.02.20 18:16:32 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 206 != 202
11.02.20 18:16:32 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 213 != 207
11.02.20 18:16:32 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 223 != 214
11.02.20 18:16:32 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 230 != 224
11.02.20 18:16:32 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 238 != 231
11.02.20 18:16:33 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 248 != 239
11.02.20 18:16:33 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 255 != 250
11.02.20 18:16:33 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 262 != 256
11.02.20 18:16:33 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 273 != 263
11.02.20 18:16:33 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 279 != 274
11.02.20 18:16:33 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 287 != 280
11.02.20 18:16:34 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 297 != 288
11.02.20 18:16:34 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 304 != 299
11.02.20 18:16:34 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 311 != 305
11.02.20 18:16:34 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 318 != 312
11.02.20 18:16:34 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 328 != 319
11.02.20 18:16:34 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 336 != 329
11.02.20 18:16:35 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 342 != 337
11.02.20 18:16:35 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 353 != 343
11.02.20 18:16:35 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 360 != 355
11.02.20 18:16:35 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 367 != 361
11.02.20 18:16:35 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 377 != 368
11.02.20 18:16:35 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 384 != 378
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 392 != 385
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 402 != 393
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 409 != 404
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 417 != 410
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 423 != 418
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 434 != 424
11.02.20 18:16:36 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 441 != 435
11.02.20 18:16:37 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 447 != 442
11.02.20 18:16:37 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 458 != 449
11.02.20 18:16:37 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 465 != 459
11.02.20 18:16:37 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 472 != 466
11.02.20 18:16:37 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 483 != 473
11.02.20 18:16:37 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 490 != 484
11.02.20 18:16:38 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 496 != 491
11.02.20 18:16:38 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 508 != 498
11.02.20 18:16:38 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 514 != 509
11.02.20 18:16:38 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 521 != 515
11.02.20 18:16:38 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 528 != 522
11.02.20 18:16:38 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 538 != 529
11.02.20 18:16:39 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 546 != 539
11.02.20 18:16:39 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 553 != 548
11.02.20 18:16:39 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 563 != 554
11.02.20 18:16:39 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 570 != 564
11.02.20 18:16:39 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 577 != 571
11.02.20 18:16:39 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 587 != 578
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 595 != 588
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 602 != 596
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 610 != 604
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 620 != 611
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 627 != 621
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 633 != 628
11.02.20 18:16:40 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 644 != 634
11.02.20 18:16:41 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 651 != 645
11.02.20 18:16:41 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 658 != 653
11.02.20 18:16:41 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 668 != 659
11.02.20 18:16:41 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 675 != 669
11.02.20 18:16:41 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 683 != 676
11.02.20 18:16:41 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 693 != 684
11.02.20 18:16:42 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 700 != 694
11.02.20 18:16:42 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 707 != 702
11.02.20 18:16:42 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 714 != 708
11.02.20 18:16:42 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 725 != 715
11.02.20 18:16:42 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 732 != 726
11.02.20 18:16:42 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 738 != 733
11.02.20 18:16:43 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 749 != 739
11.02.20 18:16:43 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 756 != 751
11.02.20 18:16:43 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 763 != 757
11.02.20 18:16:43 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 774 != 764
11.02.20 18:16:43 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 780 != 775
11.02.20 18:16:43 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 788 != 781
11.02.20 18:16:44 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 798 != 789
11.02.20 18:16:44 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 805 != 799
11.02.20 18:16:44 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 813 != 807
11.02.20 18:16:44 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 819 != 814
11.02.20 18:16:44 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 830 != 820
11.02.20 18:16:44 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 836 != 831
11.02.20 18:16:45 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 843 != 837
11.02.20 18:16:45 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 854 != 844
11.02.20 18:16:45 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 861 != 856
11.02.20 18:16:45 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 868 != 862
11.02.20 18:16:45 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 879 != 869
11.02.20 18:16:45 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 885 != 880
11.02.20 18:16:46 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 892 != 886
11.02.20 18:16:46 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 902 != 894
11.02.20 18:16:46 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 910 != 903
11.02.20 18:16:46 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 917 != 911
11.02.20 18:16:46 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 925 != 918
11.02.20 18:16:46 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 935 != 926
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 942 != 936
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 948 != 943
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 958 != 950
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 966 != 959
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 974 != 967
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 983 != 975
11.02.20 18:16:47 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 991 != 984
11.02.20 18:16:48 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 997 != 992
11.02.20 18:16:48 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1008 != 998
11.02.20 18:16:48 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1015 != 1010
11.02.20 18:16:48 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1022 != 1016
11.02.20 18:16:48 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1029 != 1023
11.02.20 18:16:48 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1040 != 1030
11.02.20 18:16:49 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1047 != 1041
11.02.20 18:16:49 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1053 != 1048
11.02.20 18:16:49 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1064 != 1055
11.02.20 18:16:49 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1071 != 1065
11.02.20 18:16:49 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1078 != 1072
11.02.20 18:16:49 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1089 != 1079
11.02.20 18:16:50 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1095 != 1090
11.02.20 18:16:50 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1103 != 1096
11.02.20 18:16:50 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1110 != 1105
11.02.20 18:16:50 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1120 != 1111
11.02.20 18:16:50 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1127 != 1121
11.02.20 18:16:50 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1134 != 1128
11.02.20 18:16:51 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1145 != 1135
11.02.20 18:16:51 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1152 != 1146
11.02.20 18:16:51 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1158 != 1153
11.02.20 18:16:51 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1170 != 1160
11.02.20 18:16:51 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1176 != 1171
11.02.20 18:16:51 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1183 != 1177
11.02.20 18:16:52 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1194 != 1184
11.02.20 18:16:52 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1201 != 1195
11.02.20 18:16:52 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1208 != 1202
11.02.20 18:16:52 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1215 != 1209
11.02.20 18:16:52 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1225 != 1217
11.02.20 18:16:52 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1232 != 1226
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1240 != 1233
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1250 != 1241
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1257 != 1251
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1265 != 1259
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1275 != 1266
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1281 != 1276
11.02.20 18:16:53 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1288 != 1282
11.02.20 18:16:54 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1299 != 1289
11.02.20 18:16:54 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1306 != 1300
11.02.20 18:16:54 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1313 != 1307
11.02.20 18:16:54 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1320 != 1315
11.02.20 18:16:54 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1331 != 1321
11.02.20 18:16:54 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1338 != 1332
11.02.20 18:16:55 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1344 != 1339
11.02.20 18:16:55 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1355 != 1345
11.02.20 18:16:55 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1362 != 1356
11.02.20 18:16:55 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1369 != 1363
11.02.20 18:16:55 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1379 != 1371
11.02.20 18:16:55 (+0000)  bluetooth  /usr/bin/bluealsa: Missing RTP packet: 1387 != 1380

It is not the same as at others, but i think it is in conjunction with the buffer. This error occured at my device, when i decreased the buffer. (I changed 1000000 to 1000) It did work again, when i deleted completely the buffer.

Hope it helps
Alex

AlexProgrammerDE commented:
@balena-ci retest

tmigone commented:
@AlexProgrammerDE this looks awesome!

Regarding your question, you can send messages between devices using the `fleet-supervisor`. For example: 
- to broadcast a message: https://github.com/balenalabs/balena-sound/blob/master/fleet-supervisor/server.js#L42
- to listen to that message: https://github.com/balenalabs/balena-sound/blob/master/fleet-supervisor/server.js#L33

You could do something like `fleetPublisher.publish('bt-command', { command: 'whatever' })`
Let me know if this helps.

AlexProgrammerDE commented:
@chrisys and @tmigone i had an idea. 🌟 What are buttons without an case? https://github.com/balenalabs/boombeastic/ has its own cases. Could anyone create some 3D models? Buttons integrated would be needed.

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@chrisys can you enable the wiki feature? I would like to add some additional info, which would be useless in the README?

AlexProgrammerDE commented:
@chrisys i am currently building a wiki in my fork: https://github.com/AlexProgrammerDE/balena-sound/wiki

AlexProgrammerDE commented:
@balena-ci retest

chrisys commented:
Hey @AlexProgrammerDE do you think this would work with the Pimoroni Pirate Audio HATs? Check out @Fiqq90 comment in #111

AlexProgrammerDE commented:
Hi @chrisys 
It should work, but which GPIO Pins are used by it? I didnt find any documentation.

AlexProgrammerDE commented:
@chrisys i just saw the docs. I will look into it.

AlexProgrammerDE commented:
So. The BCM pins are 5, 6, 16, and 20, but i have no HAT, so i cant test it.

chrisys commented:
@Fiqq90 can you test this on your Pirate audio HAT and let us know how if it works?

AlexProgrammerDE commented:
By the way @chrisys does the code work? I had it only tested with my Android phone. Does it work with IPhones?

chrisys commented:
@AlexProgrammerDE I haven't had a chance to test yet

Fiqq90 commented:
@chrisys @AlexProgrammerDE I would love to test it out, but unfortunately I am new to all this. Any guides to get me started on testing via PR? I assume that this latest update for the pimoroni has not been merged with the master that I can download?

AlexProgrammerDE commented:
@Figg90 it is very easy to use my code. I made a fork of this repository on github. You can very easy download the zip file from there and push it like before.

AlexProgrammerDE commented:
@Fiqq90 the url of my repository is: https://github.com/AlexProgrammerDE/balena-sound

AlexProgrammerDE commented:
@balena-ci retest

Fiqq90 commented:
@AlexProgrammerDE Pushed the files from your repo. Unfortunately the buttons still do not work for the pirate audio HAT. Buttons still does nothing to control spotify via bluetooth.

Fiqq90 commented:
From the log screen, volume up is still GPIO17, bluetooth button -> GPIO27, volume down -> GPIO22.

Could we set the inputs to play/pause -> GPIO5, vol down -> GPIO6, next -> GPIO16, vol up -> GPIO20?

AlexProgrammerDE commented:
@Fiqq90 i did already think of that case. I implemented it already. `VOLUME_UP_GPIO`, `VOLUME_DOWN_GPIO` and `BLUETOOTH_GPIO` are the enviroment variables for it. You can set them to the number of the GPIO. Something like `5` `6` `16` `20`. But **don´t** use a number **2-4** times!

AlexProgrammerDE commented:
@tmigone i experience big delays with multiroom. Is there any way to force the snapcast server to play sound instant?

AlexProgrammerDE commented:
@chrisys can you test it please and give me feedback. I am almost finished.

chrisys commented:
@AlexProgrammerDE I will test it but need to find the hardware to put some buttons on a Pi first

AlexProgrammerDE commented:
@chrisys it works with jumper cables too. I put many information to the Readme. That should help you to work with it.

tmigone commented:
Testing this soon too!

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@tmigone what is your experience with my code?

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@tmigone can you test this too?

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@balena-ci retest

AlexProgrammerDE commented:
@tmigone this PR is now open for about 4 months. Can we merge it soon?

AlexProgrammerDE commented:
@balena-ci retest

tmigone commented:
hey @AlexProgrammerDE sorry it took me so long, I'll work on testing this PR this week so we can have it in 2.x.

tmigone commented on bluetooth/bluetooth-control:
```diff
@@ -0,0 +1,160 @@
+if [[ ! -d /sys/class/gpio/gpio$VOLUME_UP_GPIO ]]; then
+  echo "$VOLUME_UP_GPIO" > /sys/class/gpio/export
+  sleep 1 
+fi
+echo "in" > /sys/class/gpio/gpio$VOLUME_UP_GPIO/direction
+
+# Set up bluetooth
+if [[ ! -d /sys/class/gpio/gpio$BLUETOOTH_GPIO ]]; then
+  echo "$BLUETOOTH_GPIO" > /sys/class/gpio/export
+  sleep 1 
+fi
+echo "in" > /sys/class/gpio/gpio$BLUETOOTH_GPIO/direction
+
+# Set up - volume
+if [[ ! -d /sys/class/gpio/gpio$VOLUME_DOWN_GPIO ]]; then
+  echo "$VOLUME_DOWN_GPIO" > /sys/class/gpio/export
+  sleep 1 
+fi
+echo "in" > /sys/class/gpio/gpio$VOLUME_DOWN_GPIO/direction
+
+# Set up next
+if [[ ! -d /sys/class/gpio/gpio$NEXT_GPIO ]]; then
+  echo "$NEXT_GPIO" > /sys/class/gpio/export
+  sleep 1 
+fi
+echo "in" > /sys/class/gpio/gpio$NEXT_GPIO/direction
+
+# Set up previous
+if [[ ! -d /sys/class/gpio/gpio$PREVIOUS_GPIO ]]; then
+  echo "$PREVIOUS_GPIO" > /sys/class/gpio/export
+  sleep 1 
+fi
+echo "in" > /sys/class/gpio/gpio$PREVIOUS_GPIO/direction
+
+# Set up utput gpio (there is one energy input missing)
+if [[ ! -d /sys/class/gpio/gpio$OUTPUT_GPIO ]]; then
+  echo "$OUTPUT_GPIO" > /sys/class/gpio/export
+  sleep 1 
+fi
+echo "out" > /sys/class/gpio/gpio$OUTPUT_GPIO/direction
```
We could eliminate a lot of code using a function here. Haven't tested it, but this should be ok:

```suggestion
function configure_gpio() {
  local GPIO_PIN="$1"
  local DIRECTION="$2"

  if [[ ! -d /sys/class/gpio/gpio$GPIO_PIN ]]; then
    echo "$GPIO_PIN" > /sys/class/gpio/export
    sleep 1 
  fi
  
  echo "$DIRECTION" > /sys/class/gpio/gpio$GPIO_PIN/direction
}

configure_gpio "$VOLUME_UP_GPIO" "in"
configure_gpio "$BLUETOOTH_GPIO" "in"
configure_gpio "$VOLUME_DOWN_GPIO" "in"
configure_gpio "$NEXT_GPIO" "in"
configure_gpio "$PREVIOUS_GPIO" "in"
configure_gpio "$OUTPUT_GPIO" "out"
```

tmigone commented on bluetooth/bluetooth-control:
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
+
+# Set up variables
+VOLUME_UP_GPIO=$(</usr/src/volume_up_gpio)
+
+BLUETOOTH_GPIO=$(</usr/src/bluetooth_gpio)
+
+VOLUME_DOWN_GPIO=$(</usr/src/volume_down_gpio)
+
+NEXT_GPIO=$(</usr/src/next_gpio)
+
+PREVIOUS_GPIO=$(</usr/src/previous_gpio)
+
+OUTPUT_GPIO=$(</usr/src/output_gpio)
```
As I understand it, the other script writes the values into these files, and then this scripts reads the values. 
Are the files used anywhere else? Can't we just have all done here and avoid the files alltogether?

tmigone commented on bluetooth/bluetooth-control:
```diff
@@ -0,0 +1,160 @@
+fi
+
+# Next track
+while [[ "$(cat /sys/class/gpio/gpio$NEXT_GPIO/value)" == 1 ]]; do
+  if [[ "$SKIP_STOP" == 0 ]]; then
+  printf "Playing the next track. \n"
+  dbus-send --system --dest=org.bluez --print-reply /org/bluez/hci0/dev_$MAC_PARSED/player0 org.bluez.MediaPlayer1.Next
+  sleep 0.5
+  fi
+  SKIP_STOP="1"
+done
+
+SKIP_STOP="0"
+
+# Previous track
+while [[ "$(cat /sys/class/gpio/gpio$PREVIOUS_GPIO/value)" == 1 ]]; do
+  if [[ "$SKIP_STOP" == 0 ]]; then
+  printf "Playing the previous track. \n"
+  dbus-send --system --dest=org.bluez --print-reply /org/bluez/hci0/dev_$MAC_PARSED/player0 org.bluez.MediaPlayer1.Previous
+  sleep 0.5
+  fi
+  SKIP_STOP="1"
+done
```
Same as before, if the code is nearly identical we should try and use functions to reduce code (also easier maintenance!)

AlexProgrammerDE commented:
I think this PR is too outdated. I personally don't use balena-sound anymore, so I cannot really update it. Maybe in the far future I can come back to this.
